AstSemantics.md

```diff
@@ -231,6 +239,40 @@ operator may be added. However, due to normal fragmentation, applications are
 instead expected release unused physical pages from the working set using the
 [`discard`](FutureFeatures.md#finer-grained-control-over-memory) future feature.
 
+## Tables
+
+A *table* is similar to a linear memory whose elements, instead of being bytes,
+are opaque values of a particular *table element type*. This allows the table to
+contain values like GC references, raw OS handles, or native pointers that are
+accessed by WebAssembly code indirectly through an integer index. This feature
+bridges the gap between low-level, untrusted linear memory and high-level
+opaque handles/references at the cost of a bounds-checked table indirection.
+
+The table's element type dynamically constrains the type of elements stored 
+in the table and allows engines to avoid some type checks on table use. When
+tables are mutated, any stored value is necessarily coerced (possibly trapping)
+to the element type.
+
+In the MVP, the set of operations and types for tables is limited:
+* tables may only be accessed via [`call_table`](#calls);
+* the only allowed table element types are "function" and
+  "function with signature `X`";
+* tables may not be directly mutated or resized from WebAssembly code;
+  initially this must be done through the host environment (e.g., the
+  the `WebAssembly` [JavaScript API](JS.md#webassemblytable-objects).
+
+These restrictions should be relaxed in the 
+[future](FutureFeatures.md#more-table-operators-and-types). In the MVP,
+the primary purpose of tables is to implement indirect function calls
+in C/C++, using an integer index as the pointer-to-function and the table
+to hold the array of indirectly-callable functions. The compiler may either
+choose to group functions by signature (into N tables-with-signatures)
+or to put all functions into the same table (with no signature). The former
+strategy allows the engine to eliminate dynamic signature-mismatch checks while
+the latter strategy gives each function a unique index (which may be necessary
+for C/C++ compatibility) and admits simpler and potentially more space-efficient
+[dynamic linking](DynamicLinking.md).
```

Comment:
Splitting tables also allows some form of CFI based on classy hierarchy, similar to what [LLVM currently supports](http://clang.llvm.org/docs/ControlFlowIntegrity.html#forward-edge-cfi-for-virtual-calls).